lbrynet/reflector/server/server.py

```diff
@@ -68,26 +72,101 @@ def _get_valid_response(self, response_msg):
     def handle_request(self, request_dict):
         if self.received_handshake is False:
             return self.handle_handshake(request_dict)
-        else:
+        if 'sd_blob_hash' in request_dict and 'sd_blob_size' in request_dict:
+            if not is_valid_blobhash(request_dict['sd_blob_hash']):
+                raise InvalidBlobHashError(request_dict['sd_blob_hash'])
+            return self.handle_descriptor_request(request_dict)
+        if 'blob_hash' in request_dict and 'blob_size' in request_dict:
+            if not is_valid_blobhash(request_dict['blob_hash']):
+                raise InvalidBlobHashError(request_dict['blob_hash'])
             return self.handle_normal_request(request_dict)
+        raise ReflectorRequestError("Invalid request")
 
     def handle_handshake(self, request_dict):
         log.debug('Handling handshake')
         if 'version' not in request_dict:
-            raise ValueError("Client should send version")
+            raise ReflectorRequestError("Client should send version")
         self.peer_version = int(request_dict['version'])
-        if self.peer_version != 0:
-            raise ValueError("I don't know that version!")
+        if self.peer_version not in [REFLECTOR_V1, REFLECTOR_V2]:
+            raise ReflectorClientVersionError("Unknown version: %i" % self.peer_version)
         self.received_handshake = True
-        return defer.succeed({'version': 0})
+        d = defer.succeed({'version': self.peer_version})
+        d.addCallback(self.send_response)
+        return d
+
+    def clean_up_failed_upload(self, err, blob):
+        if err.check(DownloadCanceledError):
+            err.trap(DownloadCanceledError)
+            log.warning("Failed to receive %s", blob.blob_hash[:16])
+            self.blob_manager.delete_blobs([blob.blob_hash])
+        else:
+            log.exception(err)
+        return False
+
+    def log_received_blob(self, blob, message):
+        log.info(message, blob.blob_hash[:16])
+        return True
+
+    def determine_descriptor_needed(self, sd_blob):
```

Comment:
Again, use inline callbacks.

This function is poorly named.  When I read the name, I expect that the result with be a boolean saying either yes or no that I need the descriptor.

Also, the function is too complicated.  I will yell this until I'm hoarse: functions need to do one thing.

Each block inside the if statement could be made into their own functions.

All of the callbacks added to the blob_finished_d should be in their own function.